Sources/Target/Linux/Process.cpp

```diff
@@ -245,17 +245,29 @@ ErrorCode Process::wait() {
         CHK(ptrace().getEventMessage(_pid, spawnedThreadIdData));
         auto const spawnedThreadId = static_cast<ThreadId>(spawnedThreadIdData);
 
-        int spawnedThreadStatus;
-        ThreadId const returnedThreadId =
-            ::waitpid(spawnedThreadId, &spawnedThreadStatus, WNOHANG | __WALL);
-        if (spawnedThreadId != returnedThreadId)
-          return kErrorProcessNotFound;
-        DS2LOG(Debug, "child thread tid %d %s", returnedThreadId,
-               Stringify::WaitStatus(spawnedThreadStatus));
-
-        _currentThread->step();
-        auto newThread = new Thread(this, returnedThreadId);
-        newThread->beforeResume();
+        if (_threads.find(spawnedThreadId) == _threads.end()) {
+          // If the newly cloned thread does not yet exist within the _threads
+          // vector then we have not yet seen it with waitpid(2). The most
+          // recent
+          // waitpid(2) returns a status guaranteeing a clone event did happen.
+          // Therefore, we wait and block until we see the cloned thread.
+          int spawnedThreadStatus;
+          ThreadId returnedThreadId =
+              blocking_waitpid(spawnedThreadId, &spawnedThreadStatus, __WALL);
+          if (spawnedThreadId != returnedThreadId)
+            return kErrorProcessNotFound;
+          DS2LOG(Debug, "child thread tid %d %s", returnedThreadId,
+                 Stringify::WaitStatus(spawnedThreadStatus));
+
+          _currentThread->step();
+          auto newThread = new Thread(this, returnedThreadId);
+          newThread->beforeResume();
+        } else {
+          // The new thread corresponding to this kReasonThreadSpawn was already
+          // caught by waitpid(2). Therefore, we only need to step the current
+          // thread.
+          _currentThread->step();
```

Comment:
I think that this `_currentThread->step()` could be brought out of this `else` statement along with the one above. But I find that the explanation of the logic is much clearer just repeating the `_currentThread->step()`. And given how subtle these potential race conditions are that we're working with here I figured clarity is the best option.